### PR TITLE
Move explicit specializations to the .cc file.

### PR DESCRIPTION
**Pull request by bangerth**
This is the work Dragan Nikolic sent to the mailing list a couple of days ago. I have
committed things under his name.

### CODE REVIEW

File: include/deal.II/lac/vector.templates.h [hunk 1]
```diff
@@ -304,7 +304,7 @@ void Vector<Number>::reinit (const size_type n,
     }
 
   if (omit_zeroing_entries == false)
-    *this = static_cast<Number>(0);
+    *this = Number(0);
```
Comment (kronbichler) [inline include/deal.II/lac/vector.templates.h]: Can't you just say `Number()`? This creates the identity element with respect to addition of the respective number format and should work for double, float, complex, etc. I find that clearer than casting or constructing the 'zero'...

**Comment by kronbichler** [comment]
Except for the minor style annotation, I think this is fine (not really a surprise since it has already been discussed on the mailing list).

**Comment by bangerth** [comment]
I just also pushed a changelog entry.

File: include/deal.II/lac/vector.templates.h (hunk 1 above)
Comment (bangerth) [inline include/deal.II/lac/vector.templates.h]: I suspect you can. We are using both styles numerous times in this file. Let me know which way you want it.

Comment (kronbichler) [inline include/deal.II/lac/vector.templates.h]: I prefer `Number()`, but it doesn't have to be my way. ;)

Comment (bangerth) [inline include/deal.II/lac/vector.templates.h]: I've changed it everywhere in the file.

**Pull request merged**